src/python/pants/core/goals/export.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+from dataclasses import dataclass
+from typing import Iterable, cast
+
+from pants.base.build_root import BuildRoot
+from pants.core.util_rules.distdir import DistDir
+from pants.engine.console import Console
+from pants.engine.fs import EMPTY_DIGEST, AddPrefix, Digest, MergeDigests, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import collect_rules, goal_rule
+from pants.engine.target import Targets
+from pants.engine.unions import UnionMembership, union
+from pants.util.dirutil import absolute_symlink
+from pants.util.meta import frozen_after_init
+
+
+class ExportError(Exception):
+    pass
+
+
+@union
+@dataclass(frozen=True)
+class ExportableDataRequest:
+    """A union for exportable data provided by a backend.
+
+    Subclass and install a member of this type to export data.
+    """
+
+    targets: Targets
+
+
+@dataclass(frozen=True)
+class Symlink:
+    """A symlink from link_rel_path pointing to source_path.
+
+    source_path may be absolute, or relative to the repo root.
+
+    link_rel_path is relative to the enclosing ExportableData's reldir, and will be
+    absolutized when a location for that dir is chosen.
+    """
+
+    source_path: str
+    link_rel_path: str
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class ExportableData:
+    description: str
+    # Materialize digests and create symlinks under this reldir.
+    reldir: str
+    # Materialize this digest.
+    digest: Digest
+    # Create these symlinks. Symlinks are created after the digest is materialized,
+    # so may reference files/dirs in the digest.
+    symlinks: tuple[Symlink, ...]
+
+    def __init__(
+        self,
+        description: str,
+        reldir: str,
+        *,
+        digest: Digest = EMPTY_DIGEST,
+        symlinks: Iterable[Symlink] = tuple(),
+    ):
+        self.description = description
+        self.reldir = reldir
+        self.digest = digest
+        self.symlinks = tuple(symlinks)
+
+
+class ExportSubsystem(GoalSubsystem):
+    name = "export"
```

Comment:
What's the story with `export-codegen`? Should that still exist? I wonder if this should be something like `export-ide`?

Reply:
We can probably get rid of `export-codegen` once `export` exports it. I don't want to call it `export-ide` because it may be useful for other purposes. And that is a clunky name. This can be an "export various things" goal, with IDEs being a good use-case, but perhaps not the only one.

Reply:
I'm interested to see the evolution of `export`. Seems like it will have one-or-more subcommands for various (sub-?)subsystems.... sub-goals?. 

I'm interested in possibly adding functionality to the `export` goal for exporting a best-guess `module_mapping`.

Reply:
How would this be consumed? That functionality would be awesome, but `export` may not be the best place for it.

Reply:
By humans I suppose. You're probably right, let me switch hats: ... and I've forgotten my other use-case. Ignore me for now :smile:

Reply:
Documentation! `sphinx`-generated documentation. Does `pants` support this? I don't think I saw it in the docs.

Reply:
It doesn't, although someone else asked about this recently: https://pantsbuild.slack.com/archives/C046T6T9U/p1635972032450400

Reply:
A "docgen" plugin could invoke Sphinx for Python, javadoc for Java, scaladoc for Scala and so on. But I expect it would be its own goal, not part of `export`.

Reply:
Is there a good reason why it shouldn't be a subset of `export`? If `codegen` is a subset, why shouldn't `docgen` be as well?

Somewhere else someone (was it you?) brought up trying to keep the list of goals light to not have too much in `help`, and I kinda agree with that.

Reply:
I tend to agree, that `export` could be a generic goal for getting anything out of the pants world to disk.

Reply:
Well, export is more about exporting internal information, for consumption by other tools, such as IDEs. We'd export codegen for that reason only - in normal use codegen is a build byproduct that is not visible to the end user. But since the user may want to manually inspect it, and may want to view it in their IDE, we'd export it. 

Documentation, though, is a build product, like a PEX file or a distribution or a docker image or whatever. It is an actual formal output of the system that you can request. So I don't think `export` is the right place for it, just as we don't use `export` to get compiled code, or any other build product.

Reply:
Agree with @benjyw on this one... I expect that docs meet the bar for a dedicated goal. Likewise, I think that `export` being "for anything you want to get out of pants" is probably too broad: focusing on the IDE usecase is valid (i.e., making this page as short as possible: https://www.pantsbuild.org/docs/setting-up-an-ide).

Reply:
Yeah OK I agree. But now you have me wondering if we should `package` documentation :joy:

Reply:
Good question!